File: src/ast/node/statement.ts
import type { Pass } from "../../pass/pass";
import type { Block } from "./block";
import type { Duplex } from "./duplex";
import { Node } from "./node";

export type Standalone = Statement | Duplex<any>;

export abstract class Statement extends Node<Block> {
    public abstract execute(): ExecutionResult;

    public visit(pass: Pass): void {
        pass.visitStatement(this);
    }
}

export enum ExecutionResult {
    ContinueEvaluation,
    BreakLoop,
    ContinueLoop,
    Return
}

File: src/main/kotlin/me/honkling/miniscript/parser/ast/statement/Statement.kt
package me.honkling.miniscript.parser.ast.statement

import me.honkling.miniscript.diagnostic.Location
import me.honkling.miniscript.parser.ast.Block
import me.honkling.miniscript.parser.ast.Node

abstract class Statement(location: Location?, parent: Block) : Node<Block>(location, parent) {
    abstract fun execute(): ExecutionResult
}

enum class ExecutionResult {
    ContinueExecution,
    ContinueLoop,
    BreakLoop,
    Return
}